File: 2016/04/lib/aocday/parser.ex
defmodule AOCDay.Parser do
  @crypto_capture ~r/(?<letters>\D+)(?<sector>\d+)\[(?<check>\w+)\]/

  def parse do
    {:ok, binary} = File.read("./input.txt")
    _parse(binary)
  end

  def parse(binary) do
    _parse(binary)
  end

  defp _parse(binary) do
    binary
    |> String.trim
    |> String.split("\n")
    |> Enum.map(&formatted/1)
  end

  def formatted(line) do
    captures = Regex.named_captures(@crypto_capture, line)
    sector = captures["sector"] |> String.to_integer

    %{
      code: captures["letters"],
      sector: sector,
      checksum: captures["check"]
    }
  end
end

File: 2020/5/lib/parser.rb
# frozen_string_literal: true

require_relative "seat"

class Parser
  def self.parse(input = nil)
    input ||= File.read(File.join(File.dirname(__FILE__), "../input.txt"))

    input
      .strip
      .split("\n")
      .map { |c| Seat.new(c) }
  end
end
